File: modules/sample-springMvc/src/main/java/support/PositiveLong.java
package support;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

public class PositiveLong {
    @JsonCreator
    public static PositiveLong parse(final String s) {
        final long value = Long.parseLong(s);
        if (value < 0) {
            throw new IllegalArgumentException(value + " is not positive");
        }
        return new PositiveLong(value);
    }

    private final long value;

    private PositiveLong(final long value) {
        this.value = value;
    }

    public long getValue() {
        return this.value;
    }

    @JsonValue
    public String toString() {
        return String.valueOf(getValue());
    }
}

File: modules/sample-dropwizardScala/src/main/scala/examples/support/PositiveLong.scala
package examples.support

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.core.`type`.TypeReference
import examples.server.dropwizardScala.{JacksonImplicits => SJacksonImplicits}

class PositiveLong private (val value: Long)
object PositiveLong {
  def apply(value: Long): Option[PositiveLong] = if (value >= 0) Some(new PositiveLong(value)) else None
  @JsonCreator
  def applyUnsafe(value: Long): PositiveLong = apply(value).getOrElse({ throw new IllegalArgumentException(s"$value is not positive") })
  @JsonCreator
  def apply(str: String): PositiveLong = new PositiveLong(str.toLong)

  implicit val guardrailSEncodePositiveLong: SJacksonImplicits.GuardrailEncoder[PositiveLong] = SJacksonImplicits.GuardrailEncoder.instance
  implicit val guardrailSDecodePositiveLong: SJacksonImplicits.GuardrailDecoder[PositiveLong] = SJacksonImplicits.GuardrailDecoder.instance(new TypeReference[PositiveLong] {})
}
